### ISSUE TITLE
LightBulb doesn't recognize monitors connected after it was launched

### ISSUE THREAD

**Issue by iwconfig** [Tyrrrz/LightBulb#142]
Hello,

**TLDR:** Lightbulb does not respond to anything when monitor configuration changes. When I start Lightbulb with only my primary monitor switched on, and then switch on my second monitor, it stops working completely. The primary monitors gamma and brightness does not change. The secondary monitors gamma and brightness is reset to the default, i.e. the equivalent of not running Lightbulb. I have to restart Lightbulb with both my monitors switched on for it to work as expected.

Also if I've started Lightbulb with both my monitors switched on and then switches both of them on both again at a later point, say the next daytime, the secondary monitor gamma/brightness is still set to the previous settings, the ones of the evening/night. Specific to this case is that the secondary screen actually do change when I adjust the settings.

---

I have two monitors, lets name them Bob and John. ~~Primary monitor Bob is connected via HDMI-HDMI cable and secondary monitor John via DP-HDMI cable. Both monitors are TV sets, the secondary being older and cheaper.~~

~~For some reason Windows do not reconfigure the monitor setup when I turn off my primary monitor (otherwise a noticeable black screen flicker appears to the one monitor still turned on, while the desktop/WM is reconfigured). The desktop on the secondary monitor John is therefore still extended from the primary Bob and thus not considered to be the primary in relation to the other switched off monitor (Bob in this case). If I turn off the John while Bob continues to be turned on, Windows recognize only 1 monitor, the primary Bob. I suspect that this is a hardware or configuration related issue with my TV sets, since desktop/WM reconfiguration only occur when secondary John is off. Presumably something about HDMI hot plugging.~~

So, the problem actually occurs when secondary John comes on after primary Bob, which cause Windows to reconfigure the desktop/WM. If I do not restart Lightbulb, it stops working. It does not do anything on either monitor and it does not automatically adjust neither gamma nor brightness. If I turn on the secondary in the day, with primary already on, my both my screens are on day gamma/brightness settings in the middle of the night.

In other words, Lightbulb doesn't seem to notice changes in desktop/WM configuration through multiple displays. I have to restart Lightbulb for it to work properly on both.

~~**Note though:** I can only replicate this issue if I restart Lightbulb while secondary John is off, desktop/WM is reconfigured and Windows only recognize 1 and not 2 monitors (primary Bob). Switching on secondary John later, and Lightbulb does nothing except max gamma and brightness. It does not show that max is set (i.e. max brightness even though 10% is shown), it does not respond to settings in settings. Preview, disable/enable does nothing. Only a restart solves the issue.~~

~~**But here's the thing,** I do seldom restart Lightbulb, of course. So something is happening along the way while idling. I'm guessing it just sits on hold doing nothing until the transition periods? Could the reason be because Windows is freezing the process in the background to improve performance? The same issue occur with version 1.6.3.6 or 1.6.3.4 (don't remember) as well as version 2.0.~~

It's hard to replicate the issue so I cannot guarantee that I've accurately described the symptoms, but I think so. I just notice that problems have appeared after I started using my two buddies, Bob and John. Before that, it was just Bob. :man_shrugging:

Other than that, awesome little piece software you got there pal. Thanks a lot! 😃

**Comment by Tyrrrz** [comment]
Hi.
Windows kinda sucks when it comes to reporting changes in monitor configuration. I've added a workaround feature for the next release called Gamma Polling. That force-refreshes gamma every few seconds to keep it in sync for these rare cases. It's not released yet but you can try the latest CI build: https://github.com/Tyrrrz/LightBulb/actions/runs/49690507 (look for artifacts).

**Comment by iwconfig** [comment]
I tried the portable version of the build, with no luck. Do I have to uninstall the release version before running the build even though I'm using the portable version?

Does this gamma polling also cause the change of brightness to respond on the screen?

Also, I think I've figured out why it only happens sometimes for me. It's quite simple actually. I'm guessing it is due to my second monitor being switched off when I'm rebooting my computer. heh

**Comment by Tyrrrz** [comment]
Hm, I see. So even if you change the settings around, you see no effect, during night-time?

**Comment by Tyrrrz** [comment]
I think I know what's the issue. Note to self, device contexts are not updated during the program lifetime so when monitor configuration changes, it goes out of date:

https://github.com/Tyrrrz/LightBulb/blob/91801bdae2a2901ea34f968a031ee35125e0a3d7/LightBulb/Services/GammaService.cs#L10-L12

**Comment by xcasxcursex** [comment]
I've been meaning to report this after I tested it enough to pin down the exact cause of the problem. Just commenting so that I can help test as changes are applied, and since this issue exists now I'll report my observations so far:


My personal use-case is that I have two monitors connected, my primary is the gaming monitor which I normally have disabled (eg by Win+P and selecting "Second Screen Only") and a secondary monitor (a large-screen TV) as the secondary monitor, which is what I normally use for pretty much everything but gaming (like right now while I'm chilling on the couch typing on github). When I boot the PC, only the secondary TV display is enabled. If I'm going to game, I enable the primary monitor before a reboot before gaming.

When I enable the primary monitor (I don't actually use Win+P, I have a shortcut to `C:\Windows\System32\DisplaySwitch.exe /extend`, but it's the same thing as Win+P) Then the primary monitor doesn't get the gamma change. Lightbulb just doesn't effect it at all.

What's interesting is that after this, gamma changes to the secondary monitor also don't work. A good way to see it is to simply disable lightbulb entirely (click the lightbulb icon in the top right of the UI) and the primary monitor stays at full gamma, but the secondary monitor stays at whatever it was before disabling. Also, using the 'Preview' feature' is quite interesting. Not only do gamma changes not apply to either monitor, the entire 'cycle' through the hours of the day is very stuttery. Normally, there's a smooth circle drawn and the gamma changes as it goes, but now, it jumps and skips its way through the circle, and all the way, there is no visual change to the gamma/brightness,

If I exit lightbulb at this point and restart it (or if I just restart the PC and start lightbulb), everything behaves as expected. Gamma changes are applied to both monitors, disabling works, the preview is smooth, etc etc. This aligns with what you've said above about the device contexts not being updated during runtime.

Perhaps of note - Let's say I boot as normal with only my second monitor enabled, then I enable the primary monitor -or if I just boot with both monitors enabled. If I start lightbulb once both monitors are enabled, then lightbulb will work completely as normal, even after enabling or disabling any monitor. This suggests that the monitor contexts you're using do not actually change when displays are enabled or disabled - just that the function you've posted above isn't actually getting ALL the monitor contexts as the name suggests, it's only getting the currently active monitor contexts. 

Not sure how this behaviour would differ though, between having multiple monitors with one disabled in software (as per my use-case) versus having one powered off entirely (as per @iwconfig use-case). I know that even when you power off the monitor, Windows still has references it it in the device manager, so perhaps there is some way to get a list of all of those devices? I don't know.

I've tried this with gamma polling and smoothing options (independently and together) and the behaviour is the same.

Sorry for the long post, hope some of this is helpful.

**Comment by Tyrrrz** [comment]
Thanks for extra info @xcasxcursex 
Indeed that seems to match with my understanding of what happened.
The device contexts are retrieved via [WinForms API](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.screen.allscreens?view=netframework-4.8) which wraps around [Win32 API](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumdisplaymonitors):

https://github.com/Tyrrrz/LightBulb/blob/91801bdae2a2901ea34f968a031ee35125e0a3d7/LightBulb.WindowsApi/DeviceContext.cs#L77-L89

I'm pretty sure it excludes disabled monitors. Listening to `SystemEvents.DisplaySettingsChanging` event and invalidating device contexts should likely fix this issue.

**Comment by xcasxcursex** [comment]
Ah good, I was hoping there'd be something you could hook into to detect the changes - and that looks like it'll work for both powered-on and software-enabled monitors :) As soon as you push the code I'll test it and keep you posted.

**Comment by Tyrrrz** [comment]
@xcasxcursex pushed a fix. Please try out the latest CI build when you have time.

**Comment by xcasxcursex** [comment]
Works like a charm!

**Comment by Tyrrrz** [comment]
Awesome!
